java/vector/src/main/java/org/apache/arrow/vector/VectorSchemaRoot.java

```diff
@@ -206,4 +206,18 @@ public String contentToTSVString() {
     }
     return sb.toString();
   }
+
+  /**
+   * Synchronizes the schema from the current vectors.
+   * In some cases, the schema and the actual vector structure may be different.
+   * This can be caused by a promoted writer (For details, please see
+   * {@link org.apache.arrow.vector.complex.impl.PromotableWriter}).
+   * For example, when writing different types of data to a {@link org.apache.arrow.vector.complex.ListVector}
+   * may lead to such a case.
+   * When this happens, this method should be called to bring the schema and vector structure in a synchronized state.
+   */
+  public void syncSchema() {
```

Comment:
I think we should compare old schema to new schema and return a boolean to let users know that the schema has changed.  In some cases I think you can simply resend the schema.  In the case of flight its not clear to me that you can update the schema along the way.

Not for this PR but one potential follow-up is to plumb through the schema change callbacks that exist at a lower level to this level.

Reply:
Good suggestion. 
I have updated the code to return a boolean indicating if the schema has changed. In addition, I have modified the code to automatically sync batch schema before sending it. So there is no need for the user code to call the syncSchema method. 

The idea of schema change callback seems reasonable. Let's do it in a separate issue.

Reply:
I double checked https://arrow.apache.org/docs/ipc.html and it seems to indicate oinly one schema for a stream so we might not want to make this automatic (sorry not at my computer so can't see the code exactly)

Reply:
Sounds reasonable. I have changed the code to throw an exception instead.